dagshub/data_engine/model/datasource.py

```diff
@@ -609,15 +610,19 @@ def save_dataset(self, name: str) -> "Datasource":
         copy_with_ds_assigned.load_from_dataset(dataset_name=name, change_query=False)
         return copy_with_ds_assigned
 
-    def log_to_mlflow(self, artifact_name=DEFAULT_MLFLOW_ARTIFACT_NAME):
+    def log_to_mlflow(self, artifact_name=DEFAULT_MLFLOW_ARTIFACT_NAME, run: Optional["mlflow.entities.Run"] = None):
         """
         Logs the current datasource state to MLflow as an artifact.
 
         Args:
-            artifact_name: Name of the artifact that will be stored in the MLflow run
-        """
-        run = mlflow.active_run()
-        mlflow.log_dict(self._to_dict(), artifact_name)
+            artifact_name: Name of the artifact that will be stored in the MLflow run.
+            run: MLflow run to save to. If ``None``, uses the active MLflow run.
+        """
+        if run is None:
+            run = mlflow.active_run()
+            if run is None:
+                raise RuntimeError("No mlflow run is currently active")
+        mlflow.MlflowClient().log_dict(run.info.run_id, self._to_dict(), artifact_name)
```

Comment:
OK sorry for misleading you about how this should behave, but I read up now in their docs and their default behavior is different from what I expected.
1. We should be using the "fluent" API [`mlflow.log_dict`](https://github.com/mlflow/mlflow/blob/a06e82cbcc537f7f935f6198b5b180cfc59846d6/mlflow/tracking/fluent.py#L1084), which creates an active run if one doesn't exist yet. This is the behavior in all their fluent APIs like log_metric.
2. You can still get either a string or a Run object as the parameter (or None), if it's a Run object then just take its ID

OK ?

Reply:
Oh, I don't know how I missed the `run_id` arg in that. Thank you for catching it.
I even drilled into the code to find _how_ they're logging the dict (that's where I got line 625 from), but completely glossed over the run_id in the function definition.

Reply:
wait, second opinion: I actually looked at where I got it from, and apparently I had 2.3.2 installed, and they didn't have that arg there yet. So the run_id in the fluent API is a newer addition.
It looks like the code I have here is almost a repeat of what theirs is doing (except for the part where I don't create a run automatically). So it's your call on whether we want a version cutoff here